### ISSUE TITLE
Metadata mode for torchaudio.dataset

### ISSUE THREAD

**Issue by leo19941227** [pytorch/audio#2539]
### 🚀 The feature

Hello,

Thanks for the handy tools for parsing the database!
I am wondering if it is possible to let the torchaudio.dataset classes have two modes:

1. Return the actual waveform
2. Return the waveform path

Others metadata like transcriptions and speaker labels can just remain the same.

### Motivation, pitch

Since usually we would like to organize a corpus into some standardized data format (like the kaldi's data directory), before loading the actual waveform. So we can have more flexibility on how many training data points we want to use, or some preprocessing steps which might be hard to do in __getitem__ and collate_fn. For example, In SUPERB ASV, if an utterance in VoxCeleb1 is not longer than 2 seconds after VAD, it is discarded and won't be used for training (There is no utterance in this case for the testing set). In the current dataset class implementation, I not yet come out a feasible way to realize this kind of logic since the __len__ is fixed.

Furthermore, by switching to waveform path mode, it becomes really fast to iterate over the entire dataset. Then, I can re-use the dataset classes as the corpus directory parsing helper, easily obtain all the transcription in LibriSpeech and train tokenizers, or easily obtain all the speakers in VoxCeleb1 to make a Categorical encoder. In the current implementation which returns the actual waveforms, it can become too slow to iterator over the entire dataset and is not tolerable for just preprocessing steps.

### Alternatives

No

**Comment by mthrok** [comment]
I think the proposal makes sense. Decoupling dataset traversal and audio decoding makes it flexible.

For the second part, I think that the most flexible approach is to have explicit table representation of dataset internally, and apply operations like "drop rows that do not meet this criteria" and "split the dataset into batches with similar length" on it. I get the feeling that this will also give user code opportunity to improve the overall throughput. 

The existing TorchAudio datasets started as simple adaptation of classic `torch.utils.Dataset`, which was not designed for advanced usage like these. I think it's good time for us to move away from it.

**Comment by leo19941227** [comment]
Cool!

**Comment by nateanl** [comment]
Hi @leo19941227, we can add the metadata feature under prototype. One potential solution is adding `fetch_metadata` method for each dataset. The returned item is mostly identical, only the waveform is replaced by the absolute path of the audio. Does that suit your need?

**Comment by leo19941227** [comment]
Hello @nateanl ,

Cool! I think this is nice.

**Comment by leo19941227** [comment]
Hi @carolineechen,

Thanks, it looks good to me!

**Comment by vadimkantorov** [comment]
If a new Dataset abstraction is implemented into torchaudio, please consider known problems of storing data/strings as native python objects: https://github.com/speechbrain/speechbrain/issues/872
